docs/proposals/modifying-output-image.md

```diff
@@ -0,0 +1,126 @@
+These are non-goals of this proposal.
+
+## Motivation
+
+### Goals
+
+- Begin to enable simple modifications of the output image by Build authors.
+
+### Non-Goals
+
+- Enable dynamic values or placeholders/templating (e.g., git commit SHA)
+- Automatically inject labels or annotations without user input
+- Enable modification of image layer contents (see [runtime image](https://github.com/shipwright-io/build/blob/master/docs/proposals/runtime-image.md))
+
+## Proposal
+
+Add new fields to describe output image labels and annotations:
+
+```yaml=
+apiVersion: shipwright.io/v1alpha1
+kind: Build
+metadata:
+  name: my-build-name
+spec:
+  output:
+    image: quay.io/my/image
+    credentials:
+      name: my-push-creds
+    # New fields
+    labels:
+      maintainer: team@my-company.com
+      description: "This is my cool image"
+    annotations:
+      org.opencontainers.image.url: https://my-company.com/images
+      org.opencontainers.image.source: https://github.com/org/repo
+```
+
+This will be enabled by embedding the existing `Image` struct in a new struct that 
+
+```go=
+type BuildSpec struct {
+    ...
+    Output OutputImage `json:"output"`
+}
+
+type OutputImage struct {
+    Image // embedding existing type
+    
+    // New fields
+    Labels      map[string]string `json:"labels,omitempty"`
+    Annotations map[string]string `json:"annotations,omitempty"`
+}
+```
+
+### Implementation Details/Notes/Constraints
+
+When a BuildRun executes for a Build that specifies `labels` or `annotations`, the BuildRun controller will append a new step to the end of the Tekton TaskRun that augments the image with the specified labels and annotations, and pushes to the same tag.
+
+This could be accomplished using a containerized CLI tool like [`crane`](https://github.com/google/go-containerregistry/blob/main/cmd/crane), or a purpose-built binary that Shipwright provides and packages in our releases.
+
+### Risks and Mitigations
+
+**Risk:** Automation that watches for image changes by tag will see two pushes as a result of the implementation above; one for the original push done by the specified BuildStrategy, and another immediately after when the next step modifies the image to set labels/annotations.
+
+**Mitigation:** We can document this behavior as expected.
+As a future improvement, we can make the Build's specified labels and annotations available as parameters to BuildStrategy authors, so they can pass labels/annotations to their build process directly.
+In this case the appended step will perform a no-op push, which will not trigger a subsequent image change notification.
+
+## Design Details
+
+### Test Plan
+
+And e2e test case will check that a Build that specifies labels and annotations, when run, produces an image with those values set.
+
+## Drawbacks
+
+It's not ideal that the implementation produces two image pushes.
```

Comment:
This reminds me of this old issue: [Remove image push steps from the build strategies #165](https://github.com/shipwright-io/build/issues/165)

If we would introduce a well-defined property that allows the build strategy author to define whether the strategy steps push the image (afair that's the only way that Buildpacks supports) or place it on a well-known directory (like through a parameter, like `$(params.image-directory)`), then we would have an extra step already to push the image. And this would be the same step that your proposal would be extending. Only for those strategies that already push the image, you would need to have a second push.

Reply:
Yeah I agree that would be nice. Unfortunately even if Shipwright supported this well, we'd still be at the mercy of build tooling to support writing image data to a local filesystem for us to push it later, and for BuildStrategy authors to implement that correctly.

In the meantime I can't think of a better way than to just modify the pushed image in-place. I'll mention this as another alternative and possible future improvement though.

Reply:
Would it really be an image push? Or `crane` is interacting with the container registry API and tweaking the actual image's metadata?
I mean, I think the concern (possible "drawback") we would like to highlight is the fact the image could get pulled before the metadata is changed by crane.

Reply:
It depends what you consider an "image push"... 😄  Technically speaking, `crane` would be pushing a new modified manifest and config blob (no new layers), and tagging that new manifest with the tag. This is in all ways equivalent to:

```
$ tmp=$(mktemp -d)
$ cat > ${tmp}/Dockerfile < EOF
FROM original-image
LABEL foo=bar
EOF
$ docker build -t original-image -f ${tmp}/Dockerfile .
$ docker push original-image
```

(...except `crane` doesn't require access to a Docker daemon)

Registries consider PUTing a new manifest as a "push" for rate limit and webhook-firing purposes, in either case, regardless of whether any new layer data is written.

Anyway, good point about the possible race condition, I'll mention that too.